src/github.com/matrix-org/dendrite/clientapi/auth/auth.go

```diff
@@ -172,20 +192,26 @@ func GenerateAccessToken() (string, error) {
 	return base64.RawURLEncoding.EncodeToString(b), nil
 }
 
-// extractAccessToken from a request, or return an error detailing what went wrong. The
-// error message MUST be human-readable and comprehensible to the client.
-func extractAccessToken(req *http.Request) (string, error) {
-	// cf https://github.com/matrix-org/synapse/blob/v0.19.2/synapse/api/auth.py#L631
+// extractApplicationServiceAccessToken retreives an access token from a request
```

Comment:
I separated out the `extractAccessToken` function into two so that we wouldn't need to do unnecessary checks depending on whether a request was originating from an application service or not. We also maintain the "mixing query and Header" error message by only checking for that edge case in `extractApplicationServiceAccessToken` which is always called first.